Why HAVING?
A HAVING clause in SQL specifies that an SQL SELECT statement should only return rows where aggregate values meet the specified conditions. It was added to the SQL language because the WHERE keyword could not be used with aggregate functions.

So let's check this out

In [1]:
import mysql.connector

In [2]:
cnx = mysql.connector.connect(user='@##$%%$#', password='@##$$%$#@#$%')
cursor = cnx.cursor()

In [3]:
sql0 = 'DROP DATABASE IF EXISTS mydata'

In [4]:
sql1 = 'CREATE DATABASE IF NOT EXISTS mydata'

In [5]:
sql2 = 'USE mydata'

In [6]:
sql3 = "CREATE TABLE IF NOT EXISTS person (Id INT AUTO_INCREMENT PRIMARY KEY, LastName VARCHAR(20),\
        FirstName VARCHAR(20), books INT)"

In [7]:
cursor.execute ( sql0 )

In [8]:
cursor.execute ( sql1 )

In [9]:
cursor.execute ( sql2 )

In [10]:
cursor.execute ( sql3 )

In [11]:
name = ['Tom', 'John', 'Dave', 'Spencer', 'Bob', 'Bill', 'Frank', 'Tom', 'Thomas', 'Mike', 'Michael']
last_name = ['Powell', 'Loves', 'Obi', 'Lemmy', 'Stone', 'Rommer', 'Korg', 'Abbah', 'Linton', 'Semid', 'Vard']
books = [ 10, 10, 2, 2, 4, 20, 5, 2, 10, 2, 8]

In [12]:
add_person = (
  "INSERT INTO person (LastName, FirstName, books ) "
  "VALUES (%s, %s, %s)"
)

In [13]:
data = []
i = 0
for element in name:
    data.append([last_name[i], name[i], books[i]])
    i += 1
print(data)

[['Powell', 'Tom', 10], ['Loves', 'John', 10], ['Obi', 'Dave', 2], ['Lemmy', 'Spencer', 2], ['Stone', 'Bob', 4], ['Rommer', 'Bill', 20], ['Korg', 'Frank', 5], ['Abbah', 'Tom', 2], ['Linton', 'Thomas', 10], ['Semid', 'Mike', 2], ['Vard', 'Michael', 8]]


In [14]:
for element in data:
    cursor.execute(add_person, element)

We have a table named person with columns: named Id, LastName, FirstName, and books. Let's pretend for a second that this is a simple database used in the library so each person has its individual id we know his/her name and we store information about the nuber of borrowd books. So now let's say we would like to know how many books each person borrowd. It's simple select.

In [15]:
sql14 = 'SELECT books FROM person'

In [16]:
cursor.execute ( sql14 )

In [17]:
for el in cursor:
    print(el)

(10,)
(10,)
(2,)
(2,)
(4,)
(20,)
(5,)
(2,)
(10,)
(2,)
(8,)


We see that there is some tendency that some people borrow sepcific numbers of books, like always 10 or 2 books
so let's group the results

In [18]:
sql15 = 'SELECT books FROM person GROUP BY books'

In [19]:
cursor.execute ( sql15 )
for el in cursor:
    print(el)

(10,)
(2,)
(4,)
(20,)
(5,)
(8,)


but now we don't know how many persons borrowd each numbers of books, we can use COUNT

In [20]:
sql16 = 'SELECT books, COUNT(*) how_many FROM person GROUP BY books'

In [21]:
cursor.execute ( sql16 )
for el in cursor:
    print(el)

(10, 3)
(2, 4)
(4, 1)
(20, 1)
(5, 1)
(8, 1)


let's say that we are interested only in groups with at least 3 people

In [22]:
sql17 = 'SELECT books, COUNT(*) how_many FROM person WHERE COUNT(*) >2 GROUP BY books'

In [23]:
cursor.execute ( sql17 )
for el in cursor:
    print(el)

DatabaseError: 1111 (HY000): Invalid use of group function

So we now see "1111 (HY000): Invalid use of group function"
you cannot refer to the aggregate function because groups have not yet been generated!
So to solve this you must put your group filtering condition in the having clause

In [24]:
sql18 = 'SELECT books, COUNT(*) how_many FROM person GROUP BY books HAVING COUNT(*) >2 '

In [25]:
cursor.execute ( sql18 )
for el in cursor:
    print(el)

(10, 3)
(2, 4)


So now you know! Why HAVING? A HAVING clause in SQL specifies that an SQL SELECT statement should only return rows where aggregate values meet the specified conditions. It was added to the SQL language because the WHERE keyword could not be used with aggregate functions.